### ISSUE TITLE
Is there a way to scale applications at scale?

### ISSUE THREAD

**Issue by dredknight** [Shoes3/shoes3#370]
Hello everyone,

Recently I was asked if it is possible to enable the user to resize my skillwheel application as texts are read hardly on 1920x1200 resolution.

I am aware how scaling goes regarding flows and stacks. You just change width and heights between 0.0 and 1.0 values.

**But what about the pictures?**
Lets say I have a slot 50x50 with a picture inside. if the slot size increases how can I force the picture inside to increase as well?

**What about texts?** 
 As the scale increases the text size remains the same so subjectively it gets smaller compared to slots.

**Comment by ccoupe** [comment]
I'm not aware of a zoom capability in the underlying gtk/cocoa and if there is one it may pixelate poorly.
I believe Images do scale to fit the given space in certain situation - you test it. Will they look pretty? - not always. SVG's will scale but you'd have to create them which may not be easy.  Para and Textblock friends can specify  the font size and weight as you already know.  Native widget's like buttons and edit_lines do not allow font changes - there is an existing issue for that - I think it can be done cross platform  but probably not as fast as your user wants. 

The tricky part is rewriting your application to be so dynamically resized - I expect you'll find many moments of grief and a few more shoes bugs.

**Comment by BackOrder** [comment]
I concur this assessment. You will have to recalculate everything dynamically for best flexibility or something precalculated. Either way, svg would be the easiest approach but your skill wheel contains lot of bitmaps for which you will have to provide different sizes for each resolution.

Shoes is lacking in this area and it has never been a necessity so far. It would sure be a nice feature. For now, you still have the luxury of relative placement which should at least ensure each Shoes component is about the place it should be. They will certainly need to be resized accordingly.

Relative placement, resizing according to window size, generate image sets for various resolutions.

**Comment by ccoupe** [comment]
> Shoes is lacking in this area and it has never been a necessity so far. It would sure be a nice feature.

In Tk and Java/awt and Gtk and Cocoa the there is a very flexible relative layout engines. GridBagLayout in Tk and it's the way GTk3 and Cocoa  believes you should do layout instead of the pixel counting that Shoes depends on.  How to do that in Shoes with backwards compatibility ?  Expert terror-it-tory.

**Comment by BackOrder** [comment]
> In Tk and Java/awt and Gtk and Cocoa the there is a very flexible relative layout engines. GridBagLayout in Tk and it's the way Gtk3 and Cocoa believe you should do layout instead of the pixel counting that Shoes depends on. How to do that in Shoes with backwards compatibility ? Expert terror-it-tory.

This is something we should ponder about, re: issue #369 and the future of Shoes. There are probably ways around it considering GTK3 isn't out of thin air, it does have some GTK2 compatibility. Speaking of which... it would be amazing to have [Glade](https://glade.gnome.org/) support in the future.

**Comment by ccoupe** [comment]
>  it would be amazing to have Glade support in the future.

I think ruby-gtk is the proper project for that. I don't see Shoes scaling up to that.  If you need that level of control, then you've outgrown Shoes.  Here is an interesting [starting point](https://github.com/cedlemo/ruby-gtk3-tutorial) for ruby-gtk3  - it even shows connecting  the xml defined widgets.  Once you outgrow Shoes you have choices with trade-offs - ruby-qt, ruby-wx .

It's not an easy jump from 'download and install shoes' and  write `Shoes.app {button "hello world" do para "Hello World" end} because you have to install C compilers and Ruby and Gtk/Qt/ and all the other developer things.  Here's an example - [the layouts in gtk3](https://developer.gnome.org/gtk3/stable/LayoutContainers.html). Which one does Shoes use - GtkFixed. It's been on deprecation watch for years -  As we discovered with freebsd and Gtk 3.20.

[Implementing Grid](https://developer.gnome.org/gtk3/stable/GtkGrid.html) in Shoes would be tricky - different co-ordinates but if is there is cocoa equivalent there might by a way to tell Shoes (per top level window) to use something other than GtkFixed.  Needs some pondering.

**Comment by ccoupe** [comment]
Consider `Shoes.app width: 800, height: 600, layout: :grid do ..... end`  The default would be :grid :fixed (normal shoes). Event's in non-fixed layout windows would **not** be sent to shoes_place_decide() but to the toolkits manager (via a shoes cross platform wrapper). I'm warming up to exploring this possibility - a lot of work but it doesn't break backwards compatibility. **Assuming** it can be done per window.

**Comment by BackOrder** [comment]
@ccoupe I like your ideas. It might work. The grid itself would get to be sent to shoes_place_decide() but whatever goes inside would be tied in the grid.

RE: ruby-gtk3
It's just plain gtk where you don't get to use nice visual effects Shoes can. Shoes is also easier for prototyping.

RE: Glade
Actually, I attempted to do a GUI builder in Shoes but issue #329 was one of the major obstacle. It is possible to bind a Shoes element with the move and drag'n'drop, got a working proof-of-concept. There might (or might not) also need some things to do on C side for a full Shoes UI builder.

**Comment by ccoupe** [comment]
I picked grid because it is sufficiently powerful for just about anything - GtkBox layout  is more like Shoes flow/stacks. After some code examination it would be a huge task to remove/hide  GtkFixed layout . The cocoa autolayout is even more than complicated.  My thought-goal is not to fit shoes widgets into Gtk3 Grid layout, **not** to make an new GtkFixed _thing_ to be managed by shoes_place_decide(). Either way, there are issues about the canvas behind a Shoes slot so it looks like a complete rewrite/duplication  of Shoes internals.  Interesting thought experiment but low on the TODO list

re: ruby-gtk3 
A precursor to Shoes 4  was @ashbb's  green-shoes (ruby and gtk2) so it is possible to put some Shoes _dsl love_ over Gtk. Green shoes didn't have packaging - but that's kind of worthless now days because of code signing.

**Comment by dredknight** [comment]
Alright I have few questions. 

If I pick up this GTK thing and try to implement it is it going to be of any help to Shoes? If it is not it will be a put off.

I just started a few tests with the slot ratio and I got into a huge party crasher thing.
Whenever my popup is activated all ratio block are jiggled and app.height increases with every popup activated. I think you know what I mean (app grows and a hight slider appears on the right side).

**Comment by dredknight** [comment]
hm... I got another idea. Is there a way to resize the application window ?
For example if the window specs are width: 1000, height: 700 can I do something like this:
```ruby
button "resize" do
    app.width = 1500
    app.height = 1200
end

This code does not work as intended but is it possible?

**Comment by ccoupe** [comment]
>  Is there a way to resize the application window ?

Not that I know of - but I don't now everything.  Do you have resizeable: true  in your Shoes.app call? 

Your app could read a .yaml settings file before the call to Shoes.app so the user could specify the size they want. Pain in the behind.

**Comment by BackOrder** [comment]
Assignments to **Shoes.App** are always problematic. The way Shoes is built underneath makes it difficult to track variable changes. That's why we get something weird like `set_window_title` (method) instead of just doing `app.title = "Go bigger, go home"`. 

Have you tried `self.width = 1500` and such?

**Comment by dredknight** [comment]
Thank you @ccoupe  and @BackOrder but it seems you are right shoes has issues with the main window itself. self.width = is the same as app.width = and it says undefined method for both.

Basically what I do when in doubt of what can be possible is debug(self.methods) so I can see all that is possible. There is self.width but no width=.

Making everything dynamic failed due to too many issues that come along with it:
- Popup makes other slots that are adjusted dynamically to jiggle around when it appears and disappears - background image cannot be resized/scaled properly;
- Using image instead of background fixes the above problem but makes shoes to accept the image not as background but as slot so it jiggles all dynamic slots around the image and not on top.

I refurbished the application manually for a larger size and this close the issue.
A few other ideas came to my mind to throw in more Shoes functionality inside but I will open another discussion for that.

Unless you have anything else to add I will close this thread in a day or so :).

**Comment by BackOrder** [comment]
It would be possible to implement `width=` and `height=`. The question is whether you really need it or not. Normally, applications define the window size before they open it. So why do you need to do it at runtime?

**Additional references:**
- Styles will not allow to position a window #28 
- Window manipulation methods #64

**Comment by dredknight** [comment]
My first idea was to make it resizeable so users can adjust them as they want but after some tries I saw that slots jiggle not that way I wanted. Then I decided to do a couple of size presets like the games video menus  you can choose resolution (800x600, 1024x768 etc) but this had to happen after the application was initialized because the settings menu was part of the app.

At the end I ditched this idea as well because if anyone is running 800x600 monitor I doubt he will be able to play Heroes V so he wont need the manual anyways. And then I just had another thought that 800x600 is considered so prehistorically that I do not have to bother about it so I just did a higher resolution version of the app and ditched all other efforts.

**Comment by BackOrder** [comment]
I agree with your assessment. The good news is that explaining to me gives me an idea: you should support **fullscreen**. Shoes already support `fullscreen`/`fullscreen=`. Everything should be relative to the height and width and use appropriate icons according to the resolution. It should deliver what you are hoping for without flickering caused by the dynamic resizing and such.

**Comment by dredknight** [comment]
The problem with fulscreen is that app resolution depends on screen resolution so if the screen is bigger than the app this happens :
![image](https://user-images.githubusercontent.com/12410314/30771611-73e1fde2-a054-11e7-9e95-6d58de94bc19.png)

This means I have to make the app dynamic which brings me back to the issues above.

**Comment by BackOrder** [comment]
I haven't set up everything required to build Shoes on FreeBSD yet. Perhaps @ccoupe can take a look at implementing `width=` and `height=` ? Multiple resolution is not Shoes' strong side.

**Comment by BackOrder** [comment]
How about `resize width, height` instead of (or in addition of) `width=` and `height=` ? It could also include named parameters to allow default values if anyone wants to change only width or height. The reason behind `resize` is that it makes sense to resize the window immediately while width/height may mean resize one parameter or both.

**Comment by dredknight** [comment]
What sorcery is this I do not see anything in the manual that regards "resize" command.
How does the syntax go as whatever I try says undefined command?

**Comment by BackOrder** [comment]
> What sorcery is this I do not see anything in the manual that regards "resize" command.
How does the syntax go as whatever I try says undefined command?

Haha! I am suggesting a new Shoes API to fulfill your dreams. :P 

It's relatively easy to implement.

**Comment by dredknight** [comment]
As far as it works as intended I like it :D

**Comment by BackOrder** [comment]
@ccoupe if you could take a look at the latest commit. It is working except when one wants to resize smaller than the initial window size.

Right now it does not implement named parameters.

**Comment by ccoupe** [comment]
How did those commits end up at shoes:/shoes instead of shoes3/shoes3?  Small enough to copy by hand once I figure out the osx side.

**Comment by BackOrder** [comment]
Isn't it a problem with your mailer again? Seems in order here and on github web interface. 

So the problem is that it won't resize down and below the initial window size. By the way, there is a commented out `shoes_native_app_resized` but I wasn't sure whether it should be used or not ¬ may need to be removed entirely or use new function code in `shoes_native_app_resize_window`.

**Comment by BackOrder** [comment]
Here is a little test program.

![image](https://user-images.githubusercontent.com/8844578/30811075-1368458e-a1d5-11e7-8cf5-341effc088a1.png)

![image](https://user-images.githubusercontent.com/8844578/30811098-29c6b0a4-a1d5-11e7-8f6e-9354c9a2e3f1.png)

```ruby
MAXWIDTH = 1600
MAXHEIGHT = 900

Shoes.app(width: 300, height: 100) do
    @width, @height = app.width, app.height
    stack margin: 10 do
        flow do
            slider fraction: (@width / MAXWIDTH.to_f) do |s|
                @width = (s.fraction * MAXWIDTH)
            end
            para "width", margin_left: 10
        end
        flow do
            slider fraction: (@height / MAXHEIGHT.to_f) do |s|
                @height = (s.fraction * MAXHEIGHT)
            end
            para "height", margin_left: 10
        end
    end

    animate(24) do
        if ((app.width != @width) or (app.height != @height))
            app.resize @width, @height
        end
    end
end
```

**Comment by dredknight** [comment]
Looks great :)

**Comment by ccoupe** [comment]
My browser git view must be different than yours. I don't see any pull requests on shoes3/shoes3 - the repo way up top of the us page. That's not an email issue.  I suspect we have different git origin definitions 

I was just looking at shoes_native_app_resized in cocoa..m - It should work.  Is this something we deleted the ruby glue to or just another have finished, undocumented surprise?

**Comment by ccoupe** [comment]
@BackOrder - I see what's happing with git. Your commit was pushed to shoes/shoes3/master - I found it when I tried to commit - I need to do a git pull when you say you have a pull request.  A tiny bit of a mess to clean up. You should `git pull` and rebuild and run the bugs/bug370.rb 

Why your pull requests end up on the old Shoes repo is still a mystery - how do you perform a pull request ? it may not be needed.

**Comment by BackOrder** [comment]
> Why your pull requests end up on the old Shoes repo is still a mystery - how do you perform a pull request ? it may not be needed.

It doesn't seem to be a problem on my side. I'm syncing with Shoes3 repo. I did pull before pushing. Weird.

>  commit 3be1374 : You don't want to increase beyond your or your users screen size - That could be a problem (crash?).

Actually, we can prevent that problem. GTK 2 had [gdk_screen_get_width](https://developer.gnome.org/gdk3/stable/GdkScreen.html#gdk-screen-get-width) and [gdk_screen_get_height](https://developer.gnome.org/gdk3/stable/GdkScreen.html#gdk-screen-get-height). 

GTK 3 doesn't make anything easy. They say it is deprecated but only superficially explain what one should do. Maybe something like this would do: https://stackoverflow.com/questions/43225956/how-to-get-the-size-of-the-screen-with-gtk

> Also it can misbehave if you shrink it smaller than the original creation size.

Right now it will just ignore smaller sizes. I had tracked down to hints minimum geometry in the Shoes code but they are actually set to 0. Can't get any lower. So I suspect there might be something else in Shoes code to prevent sizing down. What do you think?

**Comment by BackOrder** [comment]
Oh @ccoupe ! I forgot to ask if you could generate a beta version for @dredknight to test on? This feature should allow him to support multiple resolutions.

**Comment by ccoupe** [comment]
Beta  - [here](https://walkabout.mvmanila.com/public/shoes/shoes-3.3.4-gtk3-32.exe)

**Comment by BackOrder** [comment]
>  I think users needing to do this can heed the warning of misbehaving While the stackoverlow article provides a solution it's not apriority for me.

The solution is for out of bounds. Implementing the feature means we can also provide Shoes users access to screen information such as the current resolution. 

> It's really more of bug in the test script (setting lower bounds on the slider? ) BUT it's a test script, not a contract.
 
Your theory was interesting and I wrote some code to prove it but it turns out that's not the problem. There may however be some hidden elements with fixed with. 

```ruby
Shoes.app(width: 300, height: 100) do
     animate(25) do |n|
         if 0 == n % 100
             puts "downsizing"
             app.resize 250, 90
         elsif 50 == n % 100
             puts "upsizing"
             app.resize 450, 210
         elsif 75 == n % 100
             puts "resizing"
             app.resize 300, 100
         end
     end
 end
```

**Comment by dredknight** [comment]
I will test this tomorrow as I am on the road today :).

**Comment by dredknight** [comment]
@BackOrder  I did some initial tests and it works great :). It is true that I cannot shrink the app below the starting size though.  I will have to find a way to switch all slots and elements places and size with a single command. Not sure if it worths the time though. Anyway you never know where the teachings of Shoes are going to take you so I will keep it in mind.

Also a weird issue I noticed is that you can't have two parallel shoes running but this is probably because of the beta.

**Comment by BackOrder** [comment]
```ruby
Shoes.app do
     list_box(items: ["800x600", "1920x1200"], choose: "800x600") do             |resolution|
         @resx, @resy = resolution.text.split("x").collect { |n| n.to_i }
         @scalex, @scaley = (@resx / 800.0).ceil, (@resy / 600.0).ceil
         @b.style width: (75 + 10 * @scalex), height: (25 + 10 * @scaley)
     end
 
     stack(top: self.height / 2) do
         @b = button("hello", width: 75 + 10, height: 25 + 10) do
             alert("hello?")
         end
     end
 end
```

**Comment by dredknight** [comment]
Actually I am already working on it :P.  The only issue I currently face is that when slots styling is switched to relational setup and popup appears the height of the page content increases permanently so you get a scroll on the right. Each hover that triggers a popup adds to that height. 

![image](https://user-images.githubusercontent.com/12410314/31035996-f42101ba-a572-11e7-9cd7-002c0a4983b0.png)

**Comment by BackOrder** [comment]
It might be the mix of relational and fixed causing this issue. Not so easy to get it right. One sample that gets it right is [Shoes Console](https://github.com/Shoes3/shoes3/blob/master/lib/shoes/log.rb).

Shoes doesn't seem to provide a way to disable scrollbar on Shoes.app.

**Comment by ccoupe** [comment]
> Also a weird issue I noticed is that you can't have two parallel shoes running but this is probably because of the beta.

Probably **not** the beta. Many of the OS's  handle duplicates badly.

**Comment by dredknight** [comment]
Alright I managed to go around the hover issue but another one arose that I think will stop me from doing the resize thing. The problem is that it is not a flow but shoes functionality that is not desired.

I do not want the user to be able to resize the window manually except with the resize button/listbox I will create. This one is not possible because to be able to resize the app it should have the option ** resizable: true** which also allows the user to do whatever he wants.

**Comment by BackOrder** [comment]
Here is a small code to confirm @dredknight statement that Shoes app needs to be resizable to be resized programmatically. 

```ruby
Shoes.app(width: 250, height: 100, resizable: false) do
     button("resize") do
         app.resize 800, 800
     end
 end
```

One solution would be to implement `resizable` and `resizable=`  that would allow to enable resizing for the moment you resize programmatically and reenable it afterwards.

**Comment by BackOrder** [comment]
The latest commit fixes this issue. @dredknight you can continue your work with resizable as true until you get a new beta.

GTK makes a clear distinction between a user resizing and code resizing a window. I had initially used `gtk_window_resize` which is equivalent to a user resizing a window. GTK3 manual stipulates that using `gtk_widget_set_size_request` doesn't have the geometry constraints.

Additional references:
https://developer.gnome.org/gtk3/stable/GtkWindow.html#gtk-window-resize

**Comment by ccoupe** [comment]
Hoisted from the #372 commit comments. 
> What is your test scenario?

Linux, cmd line - any script or -c, -m - any window opening. Dependent on Gtk versions?

**Comment by BackOrder** [comment]
FreeBSD, gtk3-3.22.15_1

```
$ ./shoes -c

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69e430 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69e5d0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69e770 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69e910 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69eab0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69ec50 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69edf0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x80f69ef90 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x8145b0230 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x8145b03d0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(<unknown>:63980): Gtk-WARNING **: Allocating size to GtkLabel 0x8145b0570 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

```

**Comment by ccoupe** [comment]
our gtk3 code needs lots of version checks (#if #else #endif) to deal with my older version, you're newer and windows medium age. Their are preprocessor macros for the versions. We will also have to find/write the equivalent parts. Only going to get worse.

**Comment by BackOrder** [comment]
And so much GTK2 legacy code. It seems [gtk_window_set_geometry_hints](https://developer.gnome.org/gtk3/stable/GtkWindow.html#gtk-window-set-geometry-hints) has some parameters set to NULL since 3.20.

**Comment by ccoupe** [comment]
Not a lot of gtk2 left unless you mean the layout we use - that would be a much different issue. If I read correctly, us older gtk3's will have to supply the geometry_widget arg or go back to the code you used before.

**Comment by BackOrder** [comment]
Actually, it says that `geometry_widget` is always NULL since 3.20.

**Comment by ccoupe** [comment]
And my system warning message specifically states things don't match.  So I need it to match or otherwise tweek what works for you. 

Also, Cocoa appears amenable to resize requests  but I haven't done the work yet. PS - Hackingtosh is pretty unsatisfactory for any serious Mac work. I have one my Mac for the 10.9 dependencies. Painful. On PC's the different keyboards really becomes an issue. Not worth the effort IMO.
We should open a separate issue for getting git working for everyone.

**Comment by BackOrder** [comment]
> And my system warning message specifically states things don't match. So I need it to match or otherwise tweek what works for you.

You surely did this already but perhaps a `rake clean` might help. The warning doesn't make much sense.

> Cocoa appears amenable to resize requests

It is a fairly standard amongst Windowing APIs.

> Hackingtosh is pretty unsatisfactory for any serious Mac work. I have one my Mac for the 10.9 dependencies. Painful. On PC's the different keyboards really becomes an issue. Not worth the effort IMO.

Not much choice here. Mac mini isn't particularly expensive but need to keep my budget under control, so many other priorities. I will eventually set up a vm and see what kind of contributions I can do on macOS.

> We should open a separate issue for getting git working for everyone.

Good idea. Absolutely.

**Comment by BackOrder** [comment]
@dredknight how is it going so far?

**Comment by dredknight** [comment]
It works great. I did the hardest part with the actual wheel but after that I lost patience to edit lefts, tops, widths and heights.  I switched to some backend stuff for another app (I will show it when ready).

Here is some screenshots. Do not mind that dummy resize button that is used for testing.

This is old looks:
![image](https://user-images.githubusercontent.com/12410314/31307239-65882446-ab68-11e7-867f-2b155a1c0e6a.png)

This is new looks:
![image](https://user-images.githubusercontent.com/12410314/31307241-7a572020-ab68-11e7-86b7-8de18e3a9621.png)
